From ed978aa3b6c529731f407c84d7a01c15b93b4d03 Mon Sep 17 00:00:00 2001
From: Erwan SZYMANSKI <erwan.szymanski@st.com>
Date: Thu, 11 Sep 2025 10:22:10 +0200
Subject: [PATCH] serial: stm32: keep uart lock for tty_flip_buffer_push

Ensure that the uart port is locked whenever calling
tty_flip_buffer_push to avoid having concurrent access from
the tty framework when handling the buffer.

Fixes: 6333a4850621 ("serial: stm32: push DMA RX data before suspending")
Fixes: 3f6c02fa712b ("serial: stm32: Merge hard IRQ and threaded IRQ handling into single IRQ handler")

Change-Id: I805d2329465530f324497a07ee8b18ba2252d4eb
Signed-off-by: Erwan SZYMANSKI <erwan.szymanski@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/488889
Tested-by: Alain VOLMAT <alain.volmat@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
---
 drivers/tty/serial/stm32-usart.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/tty/serial/stm32-usart.c b/drivers/tty/serial/stm32-usart.c
index 49c8d85ad152..63b6e0b7e588 100644
--- a/drivers/tty/serial/stm32-usart.c
+++ b/drivers/tty/serial/stm32-usart.c
@@ -1153,9 +1153,9 @@ static irqreturn_t stm32_usart_interrupt(int irq, void *ptr)
 	if (stm32_usart_rx_dma_started(stm32_port) && !stm32_port->throttled) {
 		spin_lock(&port->lock);
 		size = stm32_usart_receive_chars(port, (sr & USART_SR_RTOF));
-		uart_unlock_and_check_sysrq(port);
 		if (size)
 			tty_flip_buffer_push(tport);
+		uart_unlock_and_check_sysrq(port);
 		ret = IRQ_HANDLED;
 	}
 
@@ -2464,9 +2464,9 @@ static int __maybe_unused stm32_usart_serial_en_wakeup(struct uart_port *port,
 			if (!stm32_usart_rx_dma_pause(stm32_port))
 				size += stm32_usart_receive_chars(port, true);
 			stm32_usart_rx_dma_terminate(stm32_port);
-			uart_unlock_and_check_sysrq_irqrestore(port, flags);
 			if (size)
 				tty_flip_buffer_push(tport);
+			uart_unlock_and_check_sysrq_irqrestore(port, flags);
 
 			stm32_usart_clr_bits(port, ofs->cr3, USART_CR3_DMAR);
 		}
-- 
2.53.0

